docs/templates/index_template.md

```diff
@@ -0,0 +1,130 @@
+
+The exact criteria for when this code path is unsafe are complex, and will be describe below, along with some examples. However, you can always safely use `hoist_type=true` either when the function *always* returns the same type (e.g. it transforms all inputs into `String` values) OR when the following three criteria are met:
+
+1. The type that `transformer` dispatches on is concrete.
+2. The type that `transformer` dispatches on contains no abstract types: that is, for any contained array type or dict types have, their eltypes are concrete and any contained data type has concrete `fieldtypes`.
+3. The function you pass to `Transformer` is type stable
+
+When set to true, `hoist_type=true` hashes the type of the pre-transformed object, prior to looping over the contents of the post-transformed object: its fields (for a data type) or the elements (for an array or dict type). Once the contents of the object are being looped over, the hashing of each concrete-typed elements or fields are skipped. For example, when hashing an `Array{Int}` the `Int` will only be hashed once, not once for every element.
+
+When `hoist_type=false` (the default for most functions) the type of the return value from `transformer` is hashed alongside the transformed value. This can be a lot slower, but ensures the that no unexpected hash collisions will occur.
+
+More precisely, this hoisting is only valid when one of these two criteria are satisfied:
+
+1. the pre-transformed type is sufficient to disambiguate the hash of the downstream object *values* absent their object *types*.
+2. the post-transformed types do not change unless the *caller inferred* type of the input it depends on changes
+
+The latter criteria is more stringent than type stability. A function input could have caller inferred type of `Any`, be type stable, and return either `Char` or an `Int` depending on the value of its input. Such a function would violate this second criteria.
+
+### Examples
+
+When is the pre-transformed type sufficient to disambiguate hashed values? First, many type-unstable functions should be considered unable to meet this criteria. The only time they are certain to be safe is when the values disambiguate the hash regardless of the type.
+
+For instance, the assmuptions of `hoist_type=true` would be violated by the function `x -> x < 0 : Char(0) ? Int32(0)` because the bits that are hashed downstream are identical, even though the type information that should be hashed with them are different: `Char` is a `StringType` and `Ing32` is a `NumberType`.
+
+In contrast, `x -> x < 0 : Char(1) : Int32(2)` is safe to use with `hoist_type=true`, because although the type changes, the byte sequence of the value never overlaps, regardless of the type rerturned. If you are confident the bit sequence will be unique in this way, you could safely use `hoist_type=true` even though the transformer is type unstable.
```

Comment:
```suggestion
In contrast, `x -> x < 0 : Char(1) : Int32(2)` is safe to use with `hoist_type=true`, because although the type changes, the byte sequence of the value never overlaps, regardless of the type returned. If you are confident the bit sequence will be unique in this way, you could safely use `hoist_type=true` even though the transformer is type unstable.
```